Libplanet/Store/DataModel.cs

```diff
@@ -0,0 +1,244 @@
+                    || type == typeof(Address?))
+                {
+                    throw new NotSupportedException(
+                        $"Nullable value type is not supported: {type}");
+                }
+                else
+                {
+                    property.SetValue(this, DecodeFromIValue(value, type));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Decodes a <see cref="BTypes.Dictionary"/> data into an instance
+        /// of <typeparamref name="T"/>.
+        /// </summary>
+        /// <typeparam name="T">The <c>type</c> to instantiate when decoding
+        /// <paramref name="encoded"/>.  Should be a concrete <c>type</c> inherited from
+        /// <see cref="DataModel"/>.</typeparam>
+        /// <param name="encoded">The <see cref="BTypes.Dictionary"/> instance to decode.</param>
+        /// <returns>A decoded instance of type <typeparamref name="T"/> as
+        /// <see cref="DataModel"/>.  Additional casting to <typeparamref name="T"/> may be
+        /// necessary.</returns>
+        /// <exception cref="TargetInvocationException">Thrown when failed to create an
+        /// instance of type <typeparamref name="T"/> with <paramref name="encoded"/>.
+        /// Its reason for failure is stored in thrown <see cref="TargetInvocationException"/>'s
+        /// <see cref="Exception.InnerException"/>.</exception>
+        /// <exception cref="NullReferenceException">Thrown when <paramref name="encoded"/> is
+        /// <c>null</c> or <c>null</c> reference is returned by inner instantiation.</exception>
+        public static DataModel Decode<T>(BTypes.Dictionary encoded)
+            where T : DataModel
+        {
+            if (encoded is BTypes.Dictionary e)
+            {
+                object instance = Activator.CreateInstance(typeof(T), encoded)
+                    ?? throw new NullReferenceException(
+                        $"Failed to decode {nameof(encoded)}: {encoded}");
+                return (T)instance;
+            }
+            else
+            {
+                throw new NullReferenceException(
+                    $"Argument {nameof(encoded)} cannot be null");
+            }
+        }
+
+        /// <summary>
+        /// Encodes an instance into a <see cref="BTypes.Dictionary"/>.
+        /// </summary>
+        /// <returns>An encoded <see cref="BTypes.Dictionary"/> instance.</returns>
+        /// <exception cref="NotSupportedException">Thrown when <c>nullable</c> value type property
+        /// or <c>null</c> value inside a reference type is encountered.</exception>
+        /// <exception cref="ArgumentException">Thrown when an unknown invalid property type is
+        /// encountered.</exception>
+        [Pure]
+        public BTypes.Dictionary Encode()
+        {
+            BTypes.Dictionary result = BTypes.Dictionary.Empty;
+            PropertyInfo[] properties = this.GetType().GetProperties();
```

Comment:
I guess `GetProperties()` can be called earlier in its constructor. (generally, reflection is a heavy work)

Reply:
I've given this some more thought while implementing as suggested. Depending on its usage, calling `GetProperties()` during construction might be worse for performance. For one thing, I've settled on using `ImmutableArray<T>`, `ImmutableList<T>`, and `ImmutableDictionary<K, V>` instead of `T[]`, `List<T>`, and `Dictionary<K, V>`, to suggest classes inherited from `DataModel` should be immutable (see #1943).

Of course, there is no mechanism to actually prevent game developers from bypassing this restriction by reassigning values to internally, but if a class is modeled to be immutable, which in my opinion is a better practice, with any "mutating" method returning a new instance, then `GetProperties()` will get called unnecessarily often.

Also there is an issue of taxing performance when data only needs to be read from the chain. 😶

Reply:
As per additional discussion, this is implemented via caching. See 7b2ae7ea8f11bbf35784d4756ec2a019441e723e.